relayer/processor/message_processor.go

```diff
@@ -0,0 +1,457 @@
+}
+
+// trackMessage stores the message tracker in the correct slice and index based on the type.
+func (mp *messageProcessor) trackMessage(tracker messageToTrack, i int) {
+	switch t := tracker.(type) {
+	case packetMessageToTrack:
+		mp.pktMsgs[i] = t
+	case channelMessageToTrack:
+		mp.chanMsgs[i] = t
+	case connectionMessageToTrack:
+		mp.connMsgs[i] = t
+	case clientICQMessageToTrack:
+		mp.clientICQMsgs[i] = t
+	}
+}
+
+// trackers returns all of the msg trackers for the current set of messages to be sent.
+func (mp *messageProcessor) trackers() (trackers []messageToTrack) {
+	for _, t := range mp.pktMsgs {
+		trackers = append(trackers, t)
+	}
+	for _, t := range mp.chanMsgs {
+		trackers = append(trackers, t)
+	}
+	for _, t := range mp.connMsgs {
+		trackers = append(trackers, t)
+	}
+	for _, t := range mp.clientICQMsgs {
+		trackers = append(trackers, t)
+	}
+	return trackers
+}
+
+func newMessageProcessor(
+	log *zap.Logger,
+	metrics *PrometheusMetrics,
+	memo string,
+	clientUpdateThresholdTime time.Duration,
+) *messageProcessor {
+	return &messageProcessor{
+		log:                       log,
+		metrics:                   metrics,
+		memo:                      memo,
+		clientUpdateThresholdTime: clientUpdateThresholdTime,
+	}
+}
+
+// processMessages is the entrypoint for the message processor.
+// it will assemble and send any pending messages.
+func (mp *messageProcessor) processMessages(
+	ctx context.Context,
+	messages pathEndMessages,
+	src, dst *pathEndRuntime,
+) error {
+	needsClientUpdate, err := mp.shouldUpdateClientNow(ctx, src, dst)
+	if err != nil {
+		return err
+	}
+
+	if err := mp.assembleMsgUpdateClient(ctx, src, dst); err != nil {
+		return err
+	}
+
+	mp.assembleMessages(ctx, messages, src, dst)
+
+	return mp.trackAndSendMessages(ctx, src, dst, needsClientUpdate)
+}
+
+// shouldUpdateClientNow determines if an update client message should be sent
+// even if there are no messages to be sent now. It will not be attempted if
+// there has not been enough blocks since the last client update attempt.
+// Otherwise, it will be attempted if either 2/3 of the trusting period
+// or the configured client update threshold duration has passed.
+func (mp *messageProcessor) shouldUpdateClientNow(ctx context.Context, src, dst *pathEndRuntime) (bool, error) {
+	var consensusHeightTime time.Time
+	if dst.clientState.ConsensusTime.IsZero() {
+		h, err := src.chainProvider.QueryIBCHeader(ctx, int64(dst.clientState.ConsensusHeight.RevisionHeight))
+		if err != nil {
+			return false, fmt.Errorf("failed to get header height: %w", err)
+		}
+		consensusHeightTime = time.Unix(0, int64(h.ConsensusState().GetTimestamp()))
+	} else {
+		consensusHeightTime = dst.clientState.ConsensusTime
+	}
+
+	clientUpdateThresholdMs := mp.clientUpdateThresholdTime.Milliseconds()
+
+	dst.lastClientUpdateHeightMu.Lock()
+	enoughBlocksPassed := (dst.latestBlock.Height - blocksToRetrySendAfter) > dst.lastClientUpdateHeight
```

Comment:
Because we are using a regular mutex right now only one thread can access `lastClientUpdateHeight` at a time when a lock is taken. We could utilize Go's `RWMutex` and take an `RLock` here which would allow multiple threads to have read access

Reply:
Good idea, but `shouldUpdateClientNow` always happens synchronously in the path processor control loop, so I don't think it's necessary as we only need to guard against a read happening at the same time as a write.